Allowed post-processors to be used in functor parameters

joshuahansel opened the pull request:
Closes #25012

joshuahansel committed:
Allowed post-processors to be used in functor parameters

Closes #25012

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1696991/) on 0a5f62a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25188/site)

This comment will be updated on new commits.

GiudGiud commented on framework/src/postprocessors/Postprocessor.C:
```diff
@@ -64,3 +65,101 @@ Postprocessor::getValue() const
+typename Postprocessor::DotType
+Postprocessor::evaluateDot(const ElemArg & /*elem_arg*/, const Moose::StateArg & /*state*/) const
+{
+  return 0;
```
I d rather an error for time derivatives
or a doOnce warning at least

joshuahansel replied:
Yeah, a doOnce sounds like a good idea. I think I'll do the same for the gradient.

GiudGiud replied:
no not gradient imo
a PP is assumed to have a 0 gradient, that's ok

joshuahansel commented:
Hm, I was assuming that this would give me exactly one PP in `FEProblemBase::addPostprocessor`:
```
    theWarehouse()
        .query()
        .template condition<AttribName>(name)
        .template condition<AttribSystem>("UserObject")
        .template condition<AttribThread>(tid)
        .queryInto(uos);
```
but apparently not when run with multiple threads. I thought `.condition<AttribThread>(tid)` would take care of this.

joshuahansel commented:
Nevermind, it was because of _zero_ matches, not multiple.

joshuahansel commented:
I'm struggling with `FEProblemBase::addPostprocessor` - I'm doing something wrong with respect to threads. For example, when running my new test with 2 threads:
```
functors/get_functor.test: *** ERROR ***
functors/get_functor.test: The following error occurred in the object "MOOSE Problem", of type "FEProblem".
functors/get_functor.test: 
functors/get_functor.test: No functor ever provided with name 'testpp', which was requested by 'get_pp'.
```
Do I need the line like `this->addFunctor(name, *functor, tid);` for `tid` such that nothing is returned by my query?

loganharbour commented:
Do we not want to get the most recently calculated value (as the docs say) instead of the result of `getValue()`? The most recently saved value is stored in the reporter system

GiudGiud commented:
I think getValue() returns the most recent value, stored in _value very often.
I think in the future we may want to have a boolean on the PP functor like "calculate PP functor on-the-fly" that would re-calculate it. but for right now having getValue() is great!

joshuahansel commented:
I agree, I'd think `getValue()` should always return the most recent value.

loganharbour commented:
In many cases it does not return the value computed at the last execute flag. I don't think it's appropriate here.

GiudGiud commented:
That's an issue with those PPs' implementation of getValue() though?
What would be the alternative to getValue for Josh's PR?

loganharbour commented:
> That's an issue with those PPs' implementation of getValue() though?

Sure, but we have an option to be consistent so why should we not? We should be using the same values that everyone gets from the PP interface

> What would be the alternative to getValue for Josh's PR?

https://github.com/idaholab/moose/blob/cdb4886e5078301450e157460d7f026822cfaf1a/framework/src/postprocessors/Postprocessor.C#L47

Store this as a reference in the postprocessor. Setup a const method to get it. Rather simple

GiudGiud commented:
Seems fine. Maybe getValue() should disappear then, it s a duplicate, and we force people to finalize their value on finalize()

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1702349/) on 09ee070 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>cdb488</th>
      <th colspan=4>#25188 09ee07</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.62%</td>
      <td>85.59%</td>
      <td>-0.03%</td>
      <td>43.28%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89812</td>
      <td>89843</td>
      <td>+31</td>
      <td>29</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15090</td>
      <td>15126</td>
      <td>+36</td>
      <td>38</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25188/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25188/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25188/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 43.28% is less than the suggested 90.0%



This comment will be updated on new commits.

joshuahansel commented:
> In many cases it does not return the value computed at the last execute flag.

I still don't understand how this is possible. Let's assume that `getValue()` is already const.

I think what you're proposing is fine, but I agree with Guillaume that it seems to imply that the separation of `getValue()` and `finalize()` is kind of pointless, if we're never allowed to assume that `getValue()` reflects the last executed `finalize()`.

loganharbour commented:
> I still don't understand how this is possible. Let's assume that getValue() is already const.

Doesn't matter if it's const or not. In a lot of cases, it returns a value computed by something else _at the point at which it is called_.

> I think what you're proposing is fine, but I agree with Guillaume that it seems to imply that the separation of getValue() and finalize() is kind of pointless, if we're never allowed to assume that getValue() reflects the last executed finalize().

The problem is that `finalize()` is a void function because it comes from `UserObject`. It'll never return a value. Sure - long term we can change the behavior to have someone say `_value = ...` in `finalize()`, but that's a big change. This is just an artifact of the API that was chosen. By only using the declared value everywhere (instead of `getValue()`), we can at least guarantee that we're doing the right thing even though the API is poor.

loganharbour commented:
I know that it's annoying, but my goal here is to at least future proof us to regolding and making your statement about the execute on timing in documentation be _completely_ valid

joshuahansel commented:
> by something else

Ok, thanks, this part clarifies it for me. Given this possibility, I think your suggestion is good, so I'll do this.

I probably did the whole `getValue() const` thing for nothing (my fault obviously) 😢

loganharbour committed:
Simplify current value, add docstrings, rename method

- Adds lots of documentation as to getValue vs getCurrentValue
- Removes unnecessary headers
- Stores the value at construction time instead of doing a map lookup
- Moves extraneous methods to source

refs #25012

loganharbour commented:
Good with me now on the value implementation. I added at least 20 lines of docstring comments, so I expect at least multiple change requests for those

loganharbour committed:
Simplify addFunctor for Postprocessors

joshuahansel commented on framework/include/postprocessors/Postprocessor.h:
```diff
@@ -10,39 +10,119 @@
+  /// Post-processor name
+  const std::string & _pp_name;
+
+  /// The current value, which is the Reporter value that changes when we execute UOs in the problem)
```
```suggestion
  /// The current value, which is the Reporter value that changes when we execute UOs in the problem
```

joshuahansel replied:
```suggestion
  /// The current value, which is the Reporter value that changes when we execute UOs in the problem
```

joshuahansel commented:
@loganharbour Thanks for the help. I was looking for a good way to get the reporter value as a constant reference - didn't see that the declare method gave one. As for the other changes to `FEProblemBase`, I don't really understand why I had to do what I did before, but it's what I found that I had to do.

loganharbour committed:
Properly add functors for threaded and non-threaded UOs

loganharbour committed:
Request the pointer with .get()

joshuahansel committed:
Update framework/include/postprocessors/Postprocessor.h

loganharbour committed:
Warn from the actual postprocessor

loganharbour commented:
We should probably test all of the functor calls for a postprocessor

joshuahansel committed:
Added tests for PP functor dot and gradient

Refs #25012

GiudGiud commented on framework/doc/content/syntax/Functors/index.md:
```diff
@@ -17,6 +18,11 @@ but for `Functors`, the space arguments can be an element, a point in an element
+Note that when using post-processors, instead of being evaluated on-the-fly at
+a point in space and time, the most recently computed value is used, in accordance
+with the `execute_on` for the post-processor.
+
```
add a note to the PP documentation saying it inherits the functor interface. Functions and variables doc should have a note already

GiudGiud commented on framework/include/postprocessors/Postprocessor.h:
```diff
@@ -10,39 +10,119 @@
 protected:
-  const std::string _pp_name;
+  /// MOOSE object
+  const MooseObject & _pp_moose_object;
```
umm this should probably be private? we dont want users working with this do we?

loganharbour replied:
There's no harm as it's typically the same object you're in, but I do like the idea of making it private to get rid of the clutter

joshuahansel committed:
Fixed test and addressed comments

Refs #25012

joshuahansel commented:
Comments should be addressed now.

loganharbour reviewed (commented):
I approve of the content that isn't my own.

@GiudGiud or @lindsayad please

GiudGiud commented:
Excellent. This will serve us well on the FV side as well!
